### PR TITLE
Add a new option for caching support

### PR DESCRIPTION
**Pull request by zsurocking**
Usually it should be called by code generators

### CODE REVIEW

**Comment by zsurocking** [comment]
And please suggest better API names and comments

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 1]
```diff
@@ -54,6 +54,7 @@
   private final Marshaller<ReqT> requestMarshaller;
   private final Marshaller<RespT> responseMarshaller;
   private final boolean idempotent;
+  private final boolean cacheable;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: I think the name was supposed to be "safe" (following HTTP's method semantics). We should coordinate with @makdharma.

**Comment by ejona86** [comment]
(Other than the name, this looks good)

**Comment by carl-mastrangelo** [comment]
@zsurocking would you mind also changing the commit line message to match the scheme used by other recent commits?  It should be something like:

core: add caching support to MethodDescriptor


The format is "package: _____"  where ____ completes the sentence "This change improves gRPC to".

**Comment by jhspaybar** [comment]
Is there more information somewhere on what the expected interactions are here, and what this will look like in proto files?  We at netflix have created a cacheable option that we add to rpc definitions that specify the keys/path in the request object to use as the key for example, and we then have an interceptor that can be configured with a cache provider for a request lifetime type cache, a vm wide cache, or an evcache.  I want to make sure I understand the implications with this new setting and how they might interact and/or interfere.

**Comment by makdharma** [comment]
Please see https://github.com/grpc/grpc/issues/7945 for detailed design. I think this should have no adverse effect on your use case. However it will be great if you can share more details about your use case and design.

**Comment by zsurocking** [comment]
Changed APIs to withSafe and isSafe

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 2]
```diff
@@ -234,7 +237,29 @@ public boolean isIdempotent() {
   @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1775")
   public MethodDescriptor<ReqT, RespT> withIdempotent(boolean idempotent) {
     return new MethodDescriptor<ReqT, RespT>(type, fullMethodName, requestMarshaller,
-        responseMarshaller, idempotent);
+        responseMarshaller, idempotent, safe);
+  }
+
+  /**
+   * Returns whether this method is safe.
+   */
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: This probably needs a more thorough explanation of what safety means.

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 3]
```diff
@@ -234,7 +237,29 @@ public boolean isIdempotent() {
...
+   * Setting safe on the method. Works only with unary rpcs.
+   * The method will try to fetch cached responses if it's both safe and idempotent. This is best
+   * effort and may have performance impact if the method is not desgined to be cacheable.
+   *
+   * @param safe whether the method is safe.
+   * @return a new copy of MethodDescriptor.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1775")
+  public MethodDescriptor<ReqT, RespT> withSafe(boolean safe) {
+    return new MethodDescriptor<ReqT, RespT>(type, fullMethodName, requestMarshaller,
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: If this only works with unary, maybe have a checkArgument on the Type being unary?

File: core/src/main/java/io/grpc/MethodDescriptor.java (hunk 2 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Done

File: core/src/main/java/io/grpc/MethodDescriptor.java (hunk 3 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Done

**Comment by zsurocking** [comment]
I'll also change the commit message before merging

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 4]
```diff
@@ -234,7 +239,30 @@ public boolean isIdempotent() {
...
+   * Returns whether this method is safe.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1775")
+  public boolean isSafe() {
+    return safe;
+  }
+
+  /**
+   * Setting safe on the method. Works only with unary rpcs.
+   * A safe request does nothing expect retrieval so it has no side effects on the server side.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Is this supposed to be a new paragraph? If so, add `<p>` at the beginning.

**Comment by ejona86** [comment]
@zsurocking LGTM

File: core/src/main/java/io/grpc/MethodDescriptor.java (hunk 4 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Done.

File: core/src/main/java/io/grpc/MethodDescriptor.java [hunk 5]
```diff
@@ -252,7 +252,8 @@ public boolean isSafe() {
 
   /**
    * Setting safe on the method. Works only with unary rpcs.
-   * A safe request does nothing expect retrieval so it has no side effects on the server side.
+   *
+   * <p>A safe request does nothing except retrieval so it has no side effects on the server side.
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: I had meant on the `isSafe`  That way users don't have to read the comment for `withSafe` to know what `isSafe` means.

Comment (zsurocking) [inline core/src/main/java/io/grpc/MethodDescriptor.java]: Also added the comment to isSafe

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @zsurocking isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Pull request merged**